### PR TITLE
cabana: add 0.4 and new dependency variants

### PR DESCRIPTION
**Pull request by streeve**
ping @junghans 

Adds hypre-cmake dependent package (fix #24817)

Fix. https://github.com/ECP-copa/Cabana/issues/427

### CODE REVIEW

**Comment by scottwittenburg** [comment]
@spackbot run pipeline

File: var/spack/repos/builtin/packages/cabana/package.py
Comment (junghans) [inline var/spack/repos/builtin/packages/cabana/package.py]: There is a way to pull `devices_variants` from the kokkos spackage.

File: var/spack/repos/builtin/packages/cabana/package.py
Comment (junghans) [inline var/spack/repos/builtin/packages/cabana/package.py]: We need the cmake variant, right?

File: var/spack/repos/builtin/packages/cabana/package.py
Comment (junghans) [reply var/spack/repos/builtin/packages/cabana/package.py]: Something like:
```
kokkos_pkg = spack.repo.get('kokkos')
kokkos_backends = kokkos_pkg. devices_variants
```

Comment (streeve) [reply var/spack/repos/builtin/packages/cabana/package.py]: Nice! Added

File: var/spack/repos/builtin/packages/cabana/package.py
Comment (streeve) [reply var/spack/repos/builtin/packages/cabana/package.py]: Yes, I forgot we removed non-cmake support for hypre. I switched it to `hypre-cmake` which would mean we're blocked by #24817 or other route for cmake support with #25781

**Comment by streeve** [comment]
ECP-copa/Cabana#427

File: var/spack/repos/builtin/packages/cabana/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/cabana/package.py]: Use underscores for these class-scope temporary attributes (here and below): `kokkos_backends`, `backend`, `version`, etc.  You definitely need to be careful to not accidentally define `Cabana.version`.

File: var/spack/repos/builtin/packages/cabana/package.py
Comment (junghans) [inline var/spack/repos/builtin/packages/cabana/package.py:40]: Where do these options go?

File: var/spack/repos/builtin/packages/cabana/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/cabana/package.py:47]: Do these removals mean this won't build for older versions? Or were those older versions relying on kokkos?

File: var/spack/repos/builtin/packages/cabana/package.py
Comment (streeve) [reply var/spack/repos/builtin/packages/cabana/package.py]: Fixed, thanks

File: var/spack/repos/builtin/packages/cabana/package.py
Comment (streeve) [reply var/spack/repos/builtin/packages/cabana/package.py:40]: Fixed - I now only define for the previous versions

File: var/spack/repos/builtin/packages/cabana/package.py
Comment (streeve) [reply var/spack/repos/builtin/packages/cabana/package.py:47]: Added back, my mistake. New versions rely on Kokkos, but previous do not

**Review by junghans — DISMISSED** [review]

File: var/spack/repos/builtin/packages/cabana/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/cabana/package.py]: The `CMakePackage.define_from_variant` will help here (and check to make sure the variant names are valid):
```suggestion
                options.append(self.define_from_variant('Cabana_ENABLE_' + backends[backend], backend))
```

File: var/spack/repos/builtin/packages/cabana/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/cabana/package.py]:
```suggestion
            options.append(self.define_from_variant('Cabana_ENABLE_MPI', 'mpi))
```

**Comment by streeve** [comment]
Added `hypre-cmake` from #24817 to pass CI here

**Review by junghans — DISMISSED** [review]

**Review by sethrj — CHANGES_REQUESTED** [review]
@junghans Can hypre-cmake be a separate PR? I think it's nonstandard to create a new package due to a new build system (there are various workarounds in place, and I think the most drastic is labeling the old build system as `-legacy` rather than labeling the new one as `-cmake`).

**Comment by junghans** [comment]
> @junghans Can hypre-cmake be a separate PR? I think it's nonstandard to create a new package due to a new build system (there are various workarounds in place, and I think the most drastic is labeling the old build system as `-legacy` rather than labeling the new one as `-cmake`).

See the discussion in https://github.com/spack/spack/pull/25781#issuecomment-916403334, in short, package rename isn't going to happen soon.

**Comment by junghans** [comment]
Plus we need `hypre-cmake` here to make the CI pass.

**Comment by sethrj** [comment]
> Plus we need `hypre-cmake` here to make the CI pass.

Right, but it's not uncommon to have multiple interdependent prs in Spack.

**Comment by sethrj** [comment]
@junghans if you think it's easier to keep the new hyper-cmake on this branch rather than a new one first, then that's fine and I'll review the new package. (You're both the author and reviewer of that commit so we need another set of eyes on it.)

File: var/spack/repos/builtin/packages/hypre-cmake/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/hypre-cmake/package.py]:
```suggestion
    variant('superlu_dist', default=False,
```
See https://github.com/spack/spack/pull/24631#discussion_r662798100

File: var/spack/repos/builtin/packages/hypre-cmake/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/hypre-cmake/package.py]:
```suggestion
        from_variant = self.define_from_variant
        args = [
            from_variant('HYPRE_WITH_MPI', 'mpi'),
            from_variant('HYPRE_WITH_OPENMP', 'openmp'),
            from_variant('HYPRE_WITH_BIGINT', 'int64'),
            from_variant('HYPRE_WITH_MIXEDINT', 'mixedint'),
            from_variant('HYPRE_WITH_COMPLEX', 'complex'),
            from_variant('BUILD_SHARED_LIBS', 'shared'),
            from_variant('HYPRE_WITH_DSUPERLU', 'superlu_dist'),
            from_variant('HYPRE_WITH_CUDA', 'cuda'),
            from_variant('HYPRE_ENABLE_UNIFIED_MEMORY', 'unified_memory'),
        ]
```

File: var/spack/repos/builtin/packages/hypre-cmake/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/hypre-cmake/package.py]: `unified_memory`

**Review by sethrj — CHANGES_REQUESTED** [review]

**Comment by junghans** [comment]
> @junghans if you think it's easier to keep the new hyper-cmake on this branch rather than a new one first, then that's fine and I'll review the new package. (You're both the author and reviewer of that commit so we need another set of eyes on it.)

Thanks for the review.

**Review by sethrj — DISMISSED** [review]
Thanks!

**Review by sethrj — APPROVED** [review]

**Comment by streeve** [comment]
This looks like an unrelated failure - not sure if you can selectively rerun jobs

**Comment by sethrj** [comment]
@spackbot run pipeline

**Pull request merged**
